native/zowed/worker.cpp

```diff
@@ -189,13 +447,282 @@ void WorkerPool::shutdown()
+      if (worker_state == WorkerState::Faulted)
+      {
+        replace_worker(i, "fault");
+        continue;
+      }
+
+      // If the worker has already crashed/exited, replace it
+      if (worker_state == WorkerState::Exited && !worker->is_stop_requested())
+      {
+        replace_worker(i, "unexpected exit");
+      }
+      else if (worker_state == WorkerState::Running && request_timeout.count() > 0)
+      {
+        const auto last_heartbeat = worker->get_last_heartbeat();
+        const auto now = std::chrono::steady_clock::now();
+        if (last_heartbeat != std::chrono::steady_clock::time_point::min() && now > last_heartbeat)
+        {
+          const auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(now - last_heartbeat);
+          if (elapsed > request_timeout)
+          {
+            LOG_WARN("Worker %zu exceeded request timeout (%lld ms > %lld ms); scheduling replacement", i, static_cast<long long>(elapsed.count()), static_cast<long long>(request_timeout.count()));
+            replace_worker(i, "heartbeat timeout", true);
+            continue;
+          }
+        }
+      }
+    }
+
+    // Sleep loop to avoid busy waiting on CPU
+    for (int i = 0; i < 5 && supervisor_running && !is_shutting_down; i++)
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+}
+
+void WorkerPool::replace_worker(size_t worker_index, const char *reason, bool force_detach)
+{
+  if (is_shutting_down.load())
+    return;
+
+  // Mark worker as not ready
+  {
+    std::lock_guard<std::mutex> lock(ready_mutex);
+
+    if (worker_index >= workers.size())
+    {
+      LOG_ERROR("Cannot replace worker at index %zu - out of range", worker_index);
+      return;
+    }
+
+    // Ensure vectors are sized appropriately
+    if (worker_index >= ready_list.size())
+      ready_list.resize(workers.size(), false);
+    if (worker_index >= replacement_attempts.size())
+    {
+      replacement_attempts.resize(workers.size(), 0);
+      next_replacement_allowed.resize(workers.size(), std::chrono::steady_clock::time_point::min());
+    }
+
+    // Decrement ready count if it was ready
+    if (ready_list[worker_index])
+    {
+      int32_t previous_ready = ready_count.fetch_sub(1);
+      if (previous_ready <= 0)
+      {
+        // Critical: ready count and ready list are inconsistent
+        // Reset count to 0 and abort replacement to prevent further corruption
+        ready_count.store(0, std::memory_order_release);
+        LOG_ERROR("Ready count inconsistency detected for worker %zu (was %d). Aborting replacement to prevent corruption.",
+                  worker_index, previous_ready);
+        return;
+      }
+      ready_list[worker_index] = false;
+    }
+  }
+
+  std::shared_ptr<Worker> old_worker;
+  bool max_attempts_reached = false;
+  size_t attempt_number = 0;
+  std::chrono::milliseconds applied_backoff(0);
+
+  // Check retries, set next backoff, and get old worker
+  {
```

Comment:
Thanks for adding these "arbitrary" blocks across many different areas of the code.
If my understanding is correct, this is really helpful and give us some memory back 🙏

Reply:
The scoped blocks were primarily meant for releasing lock guards where applicable, but also has that benefit 👍